Mathlib/SetTheory/Nimber/Field.lean

```diff
@@ -0,0 +1,222 @@
+            Nimber.mul_add x]
+          abel_nf at H ⊢
+          simpa only [two_zsmul, add_zero, zero_add] using H
+    · obtain h | h | h := lt_trichotomy (b + y) (b + c)
+      · have H := mul_ne_of_lt _ hx _ h
+        rw [Nimber.mul_add x, Nimber.mul_add a, Nimber.mul_add x] at H
+        abel_nf at H ⊢
+        simpa only [two_zsmul, zero_add] using H
+      · exact (hy.ne <| add_right_injective _ h).elim
+      · obtain ⟨b', hb, hb'⟩ | ⟨z, hz, hz'⟩ := exists_of_lt_add h
+        · have H := mul_ne_of_lt _ hx _ hb
+          have hb'' := add_eq_iff_eq_add.2 (add_comm b c ▸ hb')
+          rw [← hb', Nimber.mul_add a b', ← add_ne_add_left (a * y), ← add_ne_add_left (a * b),
+            ← add_ne_add_left (a * b'), ← hb'', Nimber.mul_add x, Nimber.mul_add x]
+          abel_nf at H ⊢
+          simpa only [two_zsmul, add_zero, zero_add] using H
+        · exact ((hz.trans hy).ne <| add_right_injective _ hz').elim
+termination_by (a, b, c)
+
+protected theorem add_mul (a b c : Nimber) : (a + b) * c = a * c + b * c := by
+  rw [Nimber.mul_comm, Nimber.mul_add, Nimber.mul_comm, Nimber.mul_comm b]
+
+private theorem add_ne_zero_of_lt {a b : Nimber} (h : b < a) : a + b ≠ 0 := by
+  rw [add_ne_zero_iff]
+  exact h.ne'
+
+protected theorem mul_assoc (a b c : Nimber) : a * b * c = a * (b * c) := by
+  apply le_antisymm <;> refine mul_le_of_forall_ne fun x hx y hy ↦ ?_
+  · obtain ⟨a', ha, b', hb, rfl⟩ := exists_of_lt_mul hx
+    have H : (a + a') * ((b + b') * (c + y)) ≠ 0 := by
+      apply mul_ne_zero _ (mul_ne_zero _ _) <;> apply add_ne_zero_of_lt
+      assumption'
+    simp only [Nimber.add_mul, Nimber.mul_add] at H ⊢
+    iterate 7 rw [Nimber.mul_assoc]
+    rw [← add_ne_add_left (a * (b * c))]
+    abel_nf at H ⊢
+    simpa only [two_zsmul, zero_add] using H
+  · obtain ⟨b', hb, c', hc, rfl⟩ := exists_of_lt_mul hy
+    have H : (a + x) * (b + b') * (c + c') ≠ 0 := by
+      apply mul_ne_zero (mul_ne_zero _ _) <;> apply add_ne_zero_of_lt
+      assumption'
+    simp only [Nimber.add_mul, Nimber.mul_add] at H ⊢
+    iterate 7 rw [← Nimber.mul_assoc]
+    rw [← add_ne_add_left (a * b * c)]
+    abel_nf at H ⊢
+    simpa only [two_zsmul, zero_add] using H
+termination_by (a, b, c)
+
+instance : IsCancelMulZero Nimber where
+  mul_left_cancel_of_ne_zero ha h := by
+    rw [← add_eq_zero, ← Nimber.mul_add, mul_eq_zero] at h
+    exact add_eq_zero.1 (h.resolve_left ha)
+  mul_right_cancel_of_ne_zero ha h := by
+    rw [← add_eq_zero, ← Nimber.add_mul, mul_eq_zero] at h
+    exact add_eq_zero.1 (h.resolve_right ha)
+
+protected theorem one_mul (a : Nimber) : 1 * a = a := by
+  apply le_antisymm
+  · refine mul_le_of_forall_ne fun x hx y hy ↦ ?_
+    rw [Nimber.lt_one_iff_zero] at hx
+    rw [hx, Nimber.one_mul, zero_mul, zero_mul, add_zero, zero_add]
+    exact hy.ne
+  · -- by_contra! doesn't work for whatever reason.
+    by_contra h
+    replace h := lt_of_not_le h
```

Comment:
```suggestion
  · by_contra! h
    replace h := h -- needed to remind `termination_by`
```

Reply:
@nomeata, do you have any idea what might be causing this, and whether this is expected behavior?

Reply:
Just from looking at the PR diff, not sure what might be causing this. Maybe `by_contra` abstracts over the environment in strange ways? With a #mwe I could have a close look, if this occurs more often.